### PR TITLE
Add new `Performance/OpenStruct` cop

### PR DESCRIPTION
**Pull request by xlts**
Add new `Performance/OpenStruct` cop which checks for `OpenStruct.new` calls. The suggested way to fix this issue is to replace `OpenStruct`s with static `Struct`s. The cop is disabled by default.

See following posts for reasoning/reference:
https://github.com/charliesome/charlie.bz/blob/master/posts/things-that-clear-rubys-method-cache.md#openstructs and https://mensfeld.pl/2015/04/ruby-global-method-cache-invalidation-impact-on-a-single-and-multithreaded-applications/

### CODE REVIEW

**Comment by mikegee** [comment]
OpenStruct doesn't still have this problem in Ruby 2.5. I'm not sure which version fixed it. Perhaps you could find out and apply a Ruby version condition to the application of this cop?

**Comment by xlts** [comment]
@mikegee right, it seems method cache implementations really do change between versions; I was also testing it on Ruby 2.3.7 and performance gain was substantial as well. Anyway, I updated the docs as per your suggestion

File: config/default.yml [hunk 1]
```diff
@@ -2063,6 +2063,11 @@ Performance/LstripRstrip:
   Enabled: true
   VersionAdded: '0.36'
 
+Performance/OpenStruct:
+  Description: 'Use `Struct` instead of `OpenStruct`'
```
Comment (bbatsov) [inline config/default.yml]: Here we have a missing `.`.

File: lib/rubocop/cop/performance/open_struct.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
+# frozen_string_literal: true
+
+module RuboCop
+  module Cop
+    module Performance
+      # This cop checks for `OpenStruct.new` calls.
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/open_struct.rb]: I'd expand the rationale here. Keep in mind that this gets published to the RuboCop's manual, so it should give as much details as possible to the end users.

File: config/default.yml [hunk 3]
```diff
@@ -2063,6 +2063,11 @@ Performance/LstripRstrip:
...
+  Enabled: false
+  VersionAdded: '0.61'
```
Comment (bbatsov) [inline config/default.yml]: The cop should also be marked as unsafe.

File: lib/rubocop/cop/performance/open_struct.rb [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+      #   # good
+      #   class MyClass
+      #     MyStruct = Struct.new(:my_key1, :my_key2)
+      #     def my_method
+      #       MyStruct.new('my_value1', 'my_value2')
+      #     end
+      #   end
+      #
+      class OpenStruct < Cop
+        MSG = 'For performance reasons, consider using `Struct` ' \
```
Comment (bbatsov) [inline lib/rubocop/cop/performance/open_struct.rb]: I'd reword this to `Consider using Struct over OpenStruct to optimize the performance`. (or something along those lines).

**Comment by xlts** [comment]
@bbatsov I applied your suggestions. As for the description in the cop class itself, I decided to use my original version here (i.e. the one referring to global method cache) which, in my opinion, best explains the rationale here.

**Comment by bbatsov** [comment]
Thanks!
